File: 1-Part/6-chapter/permutationsWithRepetition.py
def getPermsWithRep(chars, permLenght=0, prefix=''):
    indent = '.' * len(prefix)
    print(indent + 'Start, args=("' + chars + '", ' + str(permLenght) + ', "' + prefix + '")')
    if permLenght is None:
        permLenght = len(chars)

    if permLenght == 0:
        print(indent + 'Base case reached, returning', [prefix])
        return [prefix]

    results = []
    print(indent + 'Adding each char to prefix "' + prefix + '".')
    for char in chars:
        newPrefix = prefix + char
        results.extend(getPermsWithRep(chars, permLenght - 1, newPrefix))
    print(indent + 'Returning', results)
    return results


print('All permutations with repetition of JPB128:')
print(getPermsWithRep('JPB128', 4))

File: 1-Part/6-chapter/permutationsWithRepetition.js
function getPermsWithRep(chars, permLenght, prefix) {
    if (permLenght === undefined) {
        permLenght = chars.length
    }
    if (prefix === undefined) {
        prefix = ''
    }
    let indent = '.'.repeat(prefix.length)
    console.log(`${indent} Start args= (${chars, permLenght, prefix})`)

    if (permLenght === 0) {
        console.log(`${indent} Base case reached, returning ${[prefix]}`)
        return [prefix]
    }

    let results = []
    console.log(`${indent} Adding each char to prefix ${prefix}`)
    for (let char of chars) {
        let newPrefix = prefix + char
        results = results.concat(getPermsWithRep(chars, permLenght - 1, newPrefix))
    }
    console.log(`${indent} Returning ${results}`)
    return results
}

console.log('All permutations with repetition of JPB128:')
console.log(getPermsWithRep('ABC', 1, 'A'))